### DISCUSSION TITLE
converting from obj to glb then back to obj, problem and solution

### DISCUSSION THREAD

**Discussion by gwald** [zeux/meshoptimizer#591] [Show and tell]
I fixed my problem (#589), it turns out it was an the .mtl file that wasn't understood by gltfpack.

The test.obj/.mtl loaded fine in assimp viewer, but generated no UV's and there for, no texture.
mc.obj/.mtl worked but the assimp generated .mtl need fixing (two materials none, with a texture), probably an assimp issue.

Swapped out the .mtl with the original content, and it worked great!
Not sure how useful this is to the project or anyone needing help, but here it is :)

![image](https://github.com/zeux/meshoptimizer/assets/4924613/1c2acace-1c5d-4104-a9e5-270d404bdd3a)
Content of batch file to get it working:

echo https://manpages.ubuntu.com/manpages/impish/man1/gltfpack.1.html
echo assimp Version 5.2 -debug -shared -st (compiled from github)
echo gltfpack 0.19 (github release)
echo run.bat > res.txt

assimp info  mc.obj
gltfpack.exe -i mc.obj  -o test.glb -tu -sa -si 0.22 -kn  -km   -ke  -mm -noq  -v -r log.txt

assimp info  test.glb 
assimp export test.glb new.obj
copy correct.mtl new.mtl
assimp info  new.obj

assimp_viewer.exe new.obj

[obj2obj.zip](https://github.com/zeux/meshoptimizer/files/12328573/obj2obj.zip)

**Comment by zeux** [comment]
This was initially filed as a bug (?) but I don't think that is correct, so I've converted this to a discussion.

**Comment by gwald** [comment]
test.obj ("test.obj fails to create UVs" in the zip), viewed fine in assimp, but gltfpack didn't process the materials correctly, it was probably an incorrect spec .mtl issue.
But with assimp viewing it, I just assumed it was correct... I guess assimp is more flexible with .mtl files? not sure. But I got a different modeller and that worked fine.


I forgot to include the working mc.mtl file.
[obj2obj.zip](https://github.com/zeux/meshoptimizer/files/12329488/obj2obj.zip)

**Reply by zeux** [reply]
Ok, I understand your issue now; I think this is a bug in fast_obj, a little similar to https://github.com/thisistherk/fast_obj/pull/37 but basically usemtl in test.obj has a space after the material name, which leads to the material assignment not working for the mesh, which leads to gltfpack removing UV data because there's no actual texture on the mesh. It's a little similar to https://github.com/thisistherk/fast_obj/pull/37 looks like, but needs a different type of fix.

**Reply by zeux** [reply]
(I will submit a fix shortly)

**Reply by zeux** [reply]
https://github.com/zeux/meshoptimizer/pull/592 fixes this.
